bam_consensus.c

```diff
@@ -0,0 +1,1426 @@
+ *
+ * So pMM[] is the chance that given a call 'x' that it came from the
+ * x/x allele combination.  Similarly p_o[] is the chance that call
+ * 'x' came from a mismatch (non-x) / overcall (consensus=*) combination.
+ *
+ * Examples with observation (call) C and * follows
+ *
+ *  C | A  C  G  T  *          * | A  C  G  T  *
+ *  -----------------          -----------------
+ *  A | __ _M __ __ o_         A | uu uu uu uu um
+ *  C | _M MM _M _M oM         C | uu uu uu uu um
+ *  G | __ _M __ __ o_         G | uu uu uu uu um
+ *  T | __ _M __ __ o_         T | uu uu uu uu um
+ *  * | o_ oM o_ o_ oo         * | um um um um mm
+ *
+ * In calculation terms, the _M is half __ and half MM, similarly o_ and um.
+ *
+ * Relative weights of substitution vs overcall vs undercall are governed on a
+ * per base basis using the P_OVER and P_UNDER scores (subst is
+ * 1-P_OVER-P_UNDER).
+ *
+ * The heterozygosity weight though is a per column calculation as we're
+ * trying to model whether the column is pure or mixed. Hence this is done
+ * once via a prior and has no affect on the individual matrix cells.
+ */
+
+static void consensus_init(double p_het) {
+    int i;
+
+    for (i = -500; i <= 500; i++)
+        e_tab[i] = exp(i);
+    for (i = -500; i <= 500; i++)
+        e_tab2[i] = exp(i/10.);
+    for (i = 0; i <= 500; i++)
+        e_log[i] = log(i);
+
+    // Heterozygous locations
+    for (i = 0; i < 25; i++)
+        prior[i] = p_het / 20;
+    prior[0] = prior[6] = prior[12] = prior[18] = prior[24] = (1-p_het)/5;
+
+    lprior15[0]  = log(prior[0]);
+    lprior15[1]  = log(prior[1]*2);
+    lprior15[2]  = log(prior[2]*2);
+    lprior15[3]  = log(prior[3]*2);
+    lprior15[4]  = log(prior[4]*2);
+    lprior15[5]  = log(prior[6]);
+    lprior15[6]  = log(prior[7]*2);
+    lprior15[7]  = log(prior[8]*2);
+    lprior15[8]  = log(prior[9]*2);
+    lprior15[9]  = log(prior[12]);
+    lprior15[10] = log(prior[13]*2);
+    lprior15[11] = log(prior[14]*2);
+    lprior15[12] = log(prior[18]);
+    lprior15[13] = log(prior[19]*2);
+    lprior15[14] = log(prior[24]);
+
+
+    // Rewrite as new form
+    for (i = 1; i < 101; i++) {
+        double prob = 1 - pow(10, -i / 10.0);
+
+        // May want to multiply all these by 5 so pMM[i] becomes close
+        // to -0 for most data. This makes the sums increment very slowly,
+        // keeping bit precision in the accumulator.
+        pMM[i] = log(prob/5);
+        p__[i] = log((1-prob)/20);
+        p_M[i] = log((exp(pMM[i]) + exp(p__[i]))/2);
+    }
+
+    pMM[0] = pMM[1];
+    p__[0] = p__[1];
+    p_M[0] = p_M[1];
+}
+
+static inline double fast_exp(double y) {
+    if (y >= -50 && y <= 50)
+        return e_tab2[(int)(y*10)];
+
+    if (y < -500)
+        y = -500;
+    if (y > 500)
+        y = 500;
+
+    return e_tab[(int)y];
+}
+
+/* Taylor (deg 3) implementation of the log:
+ * http://www.flipcode.com/cgi-bin/fcarticles.cgi?show=63828
```

Comment:
This link is dead.  The new location looks like `https://www.flipcode.com/archives/Fast_log_Function.shtml`.

It's not clear what the copyright status of this example is, although given that this version works on doubles instead of floats it might not be a problem.